iso8601/parse.go

```diff
@@ -0,0 +1,106 @@
+package iso8601
+
+import (
+	"encoding/binary"
+	"time"
+	"unsafe"
+)
+
+// Parse parses an ISO8601 string, e.g. "2021-03-25T21:36:12Z".
+func Parse(input string) (time.Time, error) {
+	switch len(input) {
+	case 20:
+		b := unsafeStringToBytes(input)
+		_ = b[19] // compiler hint
+
+		t1 := binary.LittleEndian.Uint64(b)
+		t2 := binary.LittleEndian.Uint64(b[8:])
+		t3 := binary.LittleEndian.Uint32(b[16:])
+
+		// Check for valid "YYYY-MM-DDTHH:MM:SSZ" separators by masking input
+		// with "    -  -  T  :  :  Z". If separators are all valid, replace
+		// them with a '0' (0x30) byte and check all bytes are now numeric.
+		// If it's not input we can handle on the fast path, pass it off to
+		// time.Parse().
```

Comment:
Could this validation code go in https://github.com/segmentio/encoding/blob/master/iso8601/valid.go#L37 ?

Reply:
The idea is that we'd validate and parse in one step; reusing the intermediate u64's to speed things up.

I could copy the validation part across if we thought it would help `Valid`?

Reply:
Yeah I was thinking we could use the same validation logic in `Valid`, but maybe this is work for a different PR.

Reply:
I haven't added it to this PR, but we could do something like:

```diff
diff --git i/iso8601/valid.go w/iso8601/valid.go
index 2c4edec..859c3a2 100644
--- i/iso8601/valid.go
+++ w/iso8601/valid.go
@@ -35,6 +35,16 @@ const (
 // Valid check value to verify whether or not it is a valid iso8601 time
 // representation.
 func Valid(value string, flags ValidFlags) bool {
+   // Parse() has fast paths for 20 and 24 byte inputs, and is faster
+   // at validation+parsing than the routine is below for just validation.
+   if flags == Strict {
+       switch len(value) {
+       case 20, 24:
+           _, err := Parse(value)
+           return err == nil
+       }
+   }
+
    var ok bool
 
    // year
```